File: VB/TimeEditControl/TimeEditPopupControl/TimeEditPopupControlViewInfo.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Drawing
Imports DevExpress.XtraEditors
Imports DevExpress.Utils

Namespace TimeEditControl
	Public Class TimeEditPopupControlViewInfo
		Private privatePopupAppearance As AppearanceObject
		Public Property PopupAppearance() As AppearanceObject
			Get
				Return privatePopupAppearance
			End Get
			Set(ByVal value As AppearanceObject)
				privatePopupAppearance = value
			End Set
		End Property
		Private privateBounds As Rectangle
		Public Property Bounds() As Rectangle
			Get
				Return privateBounds
			End Get
			Set(ByVal value As Rectangle)
				privateBounds = value
			End Set
		End Property
		Public ownerControl As TimeEditPopupControl

		Public CellsList As New List(Of CellViewInfo)()
		Private _hotPointedCell As CellViewInfo
		Public Property HotPointedCell() As CellViewInfo
			Get
				Return _hotPointedCell
			End Get
			Set(ByVal value As CellViewInfo)
				If value Is _hotPointedCell Then
					Return
				End If
				If _hotPointedCell IsNot Nothing Then
					_hotPointedCell.isHotPointed = False
					RefreshCell(_hotPointedCell.cellBounds)
				End If
				If value IsNot Nothing AndAlso (Not value.isSelected) Then
					_hotPointedCell = value
					_hotPointedCell.isHotPointed = True
					RefreshCell(_hotPointedCell.cellBounds)
				End If
			End Set
		End Property
		Private _focusedHour As CellViewInfo
		Public Property FocusedHour() As CellViewInfo
			Get
				Return _focusedHour
			End Get
			Set(ByVal value As CellViewInfo)
				If value Is _focusedHour Then
					Return
				End If
				If _focusedHour IsNot Nothing Then
					_focusedHour.isSelected = False
					RefreshCell(_focusedHour.cellBounds)
				End If
				If value IsNot Nothing Then
					_focusedHour = value
					_focusedHour.isSelected = True
					RefreshCell(_focusedHour.cellBounds)
				End If
			End Set
		End Property
		Private _focusedMinute As CellViewInfo
		Public Property FocusedMinute() As CellViewInfo
			Get
				Return _focusedMinute
			End Get
			Set(ByVal value As CellViewInfo)
				If value Is _focusedHour Then
					Return
				End If
				If _focusedMinute IsNot Nothing Then
					_focusedMinute.isSelected = False
					RefreshCell(_focusedMinute.cellBounds)
				End If
				If value IsNot Nothing Then
					_focusedMinute = value
					_focusedMinute.isSelected = True
					RefreshCell(_focusedMinute.cellBounds)
				End If
			End Set
		End Property
		Private _format As CellViewInfo
		Public Property Format() As CellViewInfo
			Get
				Return _format
			End Get
			Set(ByVal value As CellViewInfo)
				If value Is _format Then
					Return
				End If
				If _format IsNot Nothing Then
					_format.isSelected = False
					RefreshCell(_format.cellBounds)
				End If
				If value IsNot Nothing Then
					_format = value
					_format.isSelected = True
					RefreshCell(_format.cellBounds)
				End If
			End Set
		End Property

		Public Sub New(ByVal ownerEdit As TimeEdit, ByVal timeEditPopupControl As TimeEditPopupControl)
			ownerControl = timeEditPopupControl
			UpdateAppearance(ownerEdit)
			Bounds = ownerControl.ClientRectangle
			CreateCellsList()
		End Sub

		Protected Overrides Sub Finalize()
			Clear()
		End Sub

		Public Overridable Sub CalcCurrentCells(ByVal dt As DateTime)
			Dim hour As Integer = dt.Hour
			If hour > 11 Then
				hour -= 12
				Format = CellsList(CellsList.Count - 1)
			Else
				Format = CellsList(CellsList.Count - 2)
			End If
			hour = If(hour = 0, 12, hour)
			FocusedHour = CellsList(hour - 1)

			Dim minute As Double = dt.Minute
			Dim temp As Double = minute / 5
			Dim minuteString() As String = temp.ToString().Split("."c)
			Dim minuteIndex As Integer = Integer.Parse(minuteString(0))
			If minuteString.Length > 1 Then
				Dim leftover As Integer = Integer.Parse(minuteString(1))
				If leftover > 5 Then
					minuteIndex += 1
				End If
			End If
			FocusedMinute = CellsList(12 + minuteIndex)
		End Sub

		Protected Overridable Sub UpdateAppearance(ByVal ownerEdit As TimeEdit)
			PopupAppearance = New AppearanceObject(AppearanceObject.EmptyAppearance)
			PopupAppearance.BackColor = Color.Black
			PopupAppearance.BackColor2 = Color.DarkBlue
			PopupAppearance.BorderColor = Color.LightBlue
		End Sub

		Private Sub RefreshCell(ByVal rec As Rectangle)
			If ownerControl Is Nothing Then
				Return
			End If
			ownerControl.Invalidate(rec)
		End Sub

		Protected Overridable Sub CreateCellsList()
			Const midIndent As Integer = 6
			AddCellsBlock(New Point(3, 20), CellViewInfo.CellType.Hour)
			Dim upperRightCell As CellViewInfo = CellsList(2)
			AddCellsBlock(New Point(upperRightCell.cellBounds.X + upperRightCell.cellBounds.Width + midIndent, upperRightCell.cellBounds.Y), CellViewInfo.CellType.Minute)
			AddAmPmCells()
		End Sub

		Private Sub AddAmPmCells()
			Const indent As Integer = 3
			Dim lowerLeftCell As CellViewInfo = CellsList(9)
			Dim startPoint As New Point(lowerLeftCell.cellBounds.X, lowerLeftCell.cellBounds.Y + lowerLeftCell.cellBounds.Height + indent)
			Dim cellRect As New Rectangle(startPoint, New Size(30, 20))
			CellsList.Add(New CellViewInfo(PopupAppearance.Font, "AM", cellRect, CellViewInfo.CellType.Format))
			cellRect.Location = New Point(cellRect.Location.X + CellsList(CellsList.Count - 1).cellBounds.Width, cellRect.Location.Y)
			CellsList.Add(New CellViewInfo(PopupAppearance.Font, "PM", cellRect, CellViewInfo.CellType.Format))
		End Sub

		Private Sub AddCellsBlock(ByVal startPoint As Point, ByVal Type As CellViewInfo.CellType)
			Dim r As Rectangle = Rectangle.Empty
			For i As Integer = 0 To 3
				For j As Integer = 0 To 2

					If i = 0 AndAlso j = 0 Then
						r = New Rectangle(startPoint, New Size(26, 18))
					Else
						r.Location = New Point(r.Location.X + r.Width, r.Y)
					End If
					If Type = CellViewInfo.CellType.Hour Then
						CellsList.Add(New CellViewInfo(PopupAppearance.Font, (i * 3 + j + 1).ToString(), r, Type))
					Else
						Dim x As Integer = (i * 3 + j) * 5
						Dim result As String = If(x < 10, String.Format("0{0}", x), x.ToString())
						CellsList.Add(New CellViewInfo(PopupAppearance.Font, result, r, Type))
					End If
				Next j
				r.X = startPoint.X - r.Width
				r.Y += r.Height
			Next i
		End Sub

		Private Sub Clear()
			PopupAppearance.Dispose()
			CellsList.Clear()
			Bounds = Rectangle.Empty
			ownerControl = Nothing
			HotPointedCell = Nothing
			FocusedMinute = Nothing
			FocusedHour = Nothing
			Format = Nothing

		End Sub
	End Class
End Namespace

File: CS/TimeEditControl/TimeEditPopupControl/TimeEditPopupControlViewInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Drawing;
using DevExpress.XtraEditors;
using DevExpress.Utils;

namespace TimeEditControl
{
    public class TimeEditPopupControlViewInfo
    {       
        public AppearanceObject PopupAppearance { get; set; }
        public Rectangle Bounds { get; set; }
        public TimeEditPopupControl ownerControl;

        public List<CellViewInfo> CellsList = new List<CellViewInfo>();
        private CellViewInfo _hotPointedCell;
        public CellViewInfo HotPointedCell
        {
            get
            {
                return _hotPointedCell;
            }
            set
            {
                if (value == _hotPointedCell)
                    return;
                if (_hotPointedCell != null)
                {
                    _hotPointedCell.isHotPointed = false;
                    RefreshCell(_hotPointedCell.cellBounds);
                }
                if (value != null && !value.isSelected)
                {
                    _hotPointedCell = value;
                    _hotPointedCell.isHotPointed = true;
                    RefreshCell(_hotPointedCell.cellBounds);
                }
            }
        }
        private CellViewInfo _focusedHour;
        public CellViewInfo FocusedHour
        {
            get
            {
                return _focusedHour;
            }
            set
            {
                if (value == _focusedHour)
                    return;
                if (_focusedHour != null)
                {
                    _focusedHour.isSelected = false;
                    RefreshCell(_focusedHour.cellBounds);
                }
                if (value != null)
                {
                    _focusedHour = value;
                    _focusedHour.isSelected = true;
                    RefreshCell(_focusedHour.cellBounds);
                }
            }
        }
        private CellViewInfo _focusedMinute;
        public CellViewInfo FocusedMinute
        {
            get
            {
                return _focusedMinute;
            }
            set
            {
                if (value == _focusedHour)
                    return;
                if (_focusedMinute != null)
                {
                    _focusedMinute.isSelected = false;
                    RefreshCell(_focusedMinute.cellBounds);
                }
                if (value != null)
                {
                    _focusedMinute = value;
                    _focusedMinute.isSelected = true;
                    RefreshCell(_focusedMinute.cellBounds);
                }
            }
        }
        private CellViewInfo _format;
        public CellViewInfo Format
        {
            get
            {
                return _format;
            }
            set
            {
                if (value == _format)
                    return;
                if (_format != null)
                {
                    _format.isSelected = false;
                    RefreshCell(_format.cellBounds);
                }
                if (value != null)
                {
                    _format = value;
                    _format.isSelected = true;
                    RefreshCell(_format.cellBounds);
                }
            }
        }

        public TimeEditPopupControlViewInfo(TimeEdit ownerEdit, TimeEditPopupControl timeEditPopupControl)
        {
            ownerControl = timeEditPopupControl;
            UpdateAppearance(ownerEdit);
            Bounds = ownerControl.ClientRectangle;
            CreateCellsList();
        }

        ~TimeEditPopupControlViewInfo()
        {
            Clear();
        }

        public virtual void CalcCurrentCells(DateTime dt)
        {
            int hour = dt.Hour;
            if (hour > 11)
            {
                hour -= 12;
                Format = CellsList[CellsList.Count - 1]; 
            }
            else
                Format = CellsList[CellsList.Count - 2];
            hour = hour == 0 ? 12 : hour; 
            FocusedHour = CellsList[hour - 1];

            double minute = dt.Minute;
            double temp = minute / 5;
            string[] minuteString = temp.ToString().Split('.');
            int minuteIndex = int.Parse(minuteString[0]);
            if (minuteString.Length > 1)
            {
                int leftover = int.Parse(minuteString[1]);
                if (leftover > 5)
                    minuteIndex += 1;
            }
            FocusedMinute = CellsList[12 + minuteIndex];
        }

        protected virtual void UpdateAppearance(TimeEdit ownerEdit)
        {
            PopupAppearance = new AppearanceObject(AppearanceObject.EmptyAppearance);
            PopupAppearance.BackColor = Color.Black;
            PopupAppearance.BackColor2 = Color.DarkBlue;
            PopupAppearance.BorderColor = Color.LightBlue;
        }

        private void RefreshCell(Rectangle rec)
        {
            if (ownerControl == null)
                return;
            ownerControl.Invalidate(rec);
        }

        protected virtual void CreateCellsList()
        {
            const int midIndent = 6;
            AddCellsBlock(new Point(3, 20), CellViewInfo.CellType.Hour);
            CellViewInfo upperRightCell = CellsList[2];
            AddCellsBlock(new Point(upperRightCell.cellBounds.X + upperRightCell.cellBounds.Width + midIndent, upperRightCell.cellBounds.Y), CellViewInfo.CellType.Minute);
            AddAmPmCells();
        }

        private void AddAmPmCells()
        {
            const int indent = 3;
            CellViewInfo lowerLeftCell = CellsList[9];
            Point startPoint = new Point(lowerLeftCell.cellBounds.X, lowerLeftCell.cellBounds.Y + lowerLeftCell.cellBounds.Height + indent);
            Rectangle cellRect = new Rectangle(startPoint, new Size(30, 20));
            CellsList.Add(new CellViewInfo(PopupAppearance.Font, "AM", cellRect, CellViewInfo.CellType.Format));
            cellRect.Location = new Point(cellRect.Location.X + CellsList[CellsList.Count - 1].cellBounds.Width, cellRect.Location.Y);
            CellsList.Add(new CellViewInfo(PopupAppearance.Font, "PM", cellRect, CellViewInfo.CellType.Format));
        }

        private void AddCellsBlock(Point startPoint, CellViewInfo.CellType Type)
        {
            Rectangle r = Rectangle.Empty;
            for (int i = 0; i < 4; i++)
            {
                for (int j = 0; j < 3; j++)
                {

                    if (i == 0 && j == 0)
                        r = new Rectangle(startPoint, new Size(26, 18));
                    else
                    {
                        r.Location = new Point(r.Location.X + r.Width, r.Y);
                    }
                    if (Type == CellViewInfo.CellType.Hour)
                        CellsList.Add(new CellViewInfo(PopupAppearance.Font, (i * 3 + j + 1).ToString(), r, Type));
                    else
                    {
                        int x = (i * 3 + j) * 5;
                        string result = x < 10 ? string.Format("0{0}", x) : x.ToString();
                        CellsList.Add(new CellViewInfo(PopupAppearance.Font, result, r, Type));
                    }
                }
                r.X = startPoint.X - r.Width;
                r.Y += r.Height;
            }
        }

        private void Clear()
        {            
            PopupAppearance.Dispose();
            CellsList.Clear();
            Bounds = Rectangle.Empty;
            ownerControl = null;
            HotPointedCell = null;
            FocusedMinute = null;
            FocusedHour = null;
            Format = null;

        }
    }
}
